burn-dataset/src/vision/image_folder.rs

```diff
@@ -0,0 +1,382 @@
+    /// Convert a raw image dataset item (path-like) to a 3D image array with a target label.
+    fn map(&self, item: &ImageDatasetItemRaw) -> ImageDatasetItem {
+        // Map class string to label id
+        let label = self.classes.get(&item.target).unwrap();
+
+        // Load image from disk
+        let image = image::open(&item.image_path).unwrap();
+
+        // Image as Vec<PixelDepth>
+        let img_vec = match image.color() {
+            ColorType::L8 => image
+                .into_luma8()
+                .iter()
+                .map(|&x| PixelDepth::U8(x))
+                .collect(),
+            ColorType::La8 => image
+                .into_luma_alpha8()
+                .iter()
+                .map(|&x| PixelDepth::U8(x))
+                .collect(),
+            ColorType::L16 => image
+                .into_luma16()
+                .iter()
+                .map(|&x| PixelDepth::U16(x))
+                .collect(),
+            ColorType::La16 => image
+                .into_luma_alpha16()
+                .iter()
+                .map(|&x| PixelDepth::U16(x))
+                .collect(),
+            ColorType::Rgb8 => image
+                .into_rgb8()
+                .iter()
+                .map(|&x| PixelDepth::U8(x))
+                .collect(),
+            ColorType::Rgba8 => image
+                .into_rgba8()
+                .iter()
+                .map(|&x| PixelDepth::U8(x))
+                .collect(),
+            ColorType::Rgb16 => image
+                .into_rgb16()
+                .iter()
+                .map(|&x| PixelDepth::U16(x))
+                .collect(),
+            ColorType::Rgba16 => image
+                .into_rgba16()
+                .iter()
+                .map(|&x| PixelDepth::U16(x))
+                .collect(),
+            ColorType::Rgb32F => image
+                .into_rgb32f()
+                .iter()
+                .map(|&x| PixelDepth::F32(x))
+                .collect(),
+            ColorType::Rgba32F => image
+                .into_rgba32f()
+                .iter()
+                .map(|&x| PixelDepth::F32(x))
+                .collect(),
+            _ => panic!("Unrecognized image color type"),
+        };
+
+        ImageDatasetItem {
+            image: img_vec,
+            target: ImageTarget::Label(*label),
+        }
+    }
+}
+
+type ClassificationDatasetMapper = MapperDataset<
+    InMemDataset<ImageDatasetItemRaw>,
+    PathToImageClassificationItem,
+    ImageDatasetItemRaw,
+>;
+
+/// A generic dataset to load classification images from disk.
+pub struct ImageFolderDataset {
+    dataset: ClassificationDatasetMapper,
+}
+
+impl Dataset<ImageDatasetItem> for ImageFolderDataset {
+    fn get(&self, index: usize) -> Option<ImageDatasetItem> {
+        self.dataset.get(index)
+    }
+
+    fn len(&self) -> usize {
+        self.dataset.len()
+    }
+}
+
+impl ImageFolderDataset {
+    /// Create an image classification dataset from the root folder.
```

Comment:
Since the dataset type can be used for other things than classification, should we rename the doc or the method? If the method is only valid for classification, I would rename to `new_classification`, otherwise just update the doc.

Reply:
Yes! Will rename to `new_classification` since we only implement loading classification labels at this time.